Create common-fettec-onewire.rst (#3335)
ArduPilot/ardupilot_wiki: common/source/docs/common-fettec-onewire.rst

- Input voltage (V)
- Current draw (A)
- Power consumption (W)
- Temperature (°C)

This information allows the autopilot to:

- dynamicaly change the center frequency of the notch filters used to reduce frame vibration noise in the gyros
- log the status of each ESC to the SDCard or internal Flash, for post flight analysis
- send the status of each ESC to the Ground Station or companion computer for real-time monitoring

Connecting
==========

The protocol supports up-to 25 ESCs, but the ArduPilot implementation of it limits this to 12.
You can connect all ESCs to a single serial port using only a single signal wire from the autopilot's UART TX pin to the ESC's TLM pin and one GND wire.
If you are using 4in1 ESCs you can connect up-to 3 of them in this fashion.

TODO: add a figure here.

FullDuplex with Diode
=====================
If halfduplex is not available or working there is a workaround with a diode using the normal full duplex.
This is also working with other "non-halfduplex" Controller like ESP32.

The TLM-Pin (ESC) and RX (FC) connects to the anode (black side) of the diode.
TX (FC) connects to cathode (white striped side) of the diode.

 (TX --|<-- RX, TLM)

> Connection and Halfduplex with STM

Author: @teezeefpv please comment on this

> Connection and Halfduplex with STM

Reviewer: this section need to be removed as it will only confuse people that just want to plug their ESC.
And the part of the code means nothing without knowing what environment we are using (FreeRTOS? CubeMX?)

==================================

To have reliable 2Mbit/s Baudrate the GPIO should be set to PushPull.
On STM it is a special mode that is initialized like:

gpioInit.mode = LL_GPIO_MODE_ALTERNATE;
gpioInit.Pull = LL_GPIO_MODE_ALTERNATE;
gpio.Pull = LL_GPIO_PULL_UP;
gpioInit.OutputType = LL_GPIO_OUTPUT_PUSHPULL;
gpioInit.Alternate = LL_GPIO_AF_1;

Configuring
===========

These parameters have to be changed (in MissionPlanner under CONFIG -> Full Parameter List).

> To configure it you need to set the :ref:`SERIALx_PROTOCOL <common-serial-options#serialx-protocol>` parameter to `FETtec_OneWire` (38) and :ref:`SERIALx_OPTIONS <common-serial-options#serialx-options-parameter>` parameter to `HalfDuplex` (4), where x is the number of the serial port you are using.

Author: like I said, I looked around and did not find any.

> To configure it you need to set the :ref:`SERIALx_PROTOCOL <common-serial-options#serialx-protocol>` parameter to `FETtec_OneWire` (38) and :ref:`SERIALx_OPTIONS <common-serial-options#serialx-options-parameter>` parameter to `HalfDuplex` (4), where x is the number of the serial port you are using.

Reviewer 2: I will fix and push

For example, a serial 2 connection would need:

```

> SERIAL2_PROTOCOL = 38

Reviewer 2: these need links

Author: done

SERIAL2_OPTIONS 160 # full duplex - Use with diode
SERIAL2_OPTIONS 164 # half duplex - Use without diode -> TX from FC to TLM of FETtec ESC
```

> Change SERVO*_FUNCTION to 33-38 according to the needed motor order e.g.:

Reviewer 2: we use SERVOx_FUNCTION ...sometimes SERVOn_FUNCTION, but I am not a fan of that

Author: done

```

> MOT_PWM_MAX 2000

Reviewer 2: these are not linked...not sure what format they are in to get this highlighting

MOT_PWM_MIN 1000
SERVO1_FUNCTION 33
SERVO2_FUNCTION 34
SERVO3_FUNCTION 35
SERVO4_FUNCTION 36
```

Then you need to reboot the autopilot.

> After that you must set the channels you want to use in the SERVO_FTW_MASK parameter, and again restart your autopilot.

Reviewer 2: param not linked

> For example, a quadcopter using the first four motors would need:

Reviewer 2: ```suggestion
For example, a quadcopter using the first four motors would need the first four bits of the bitmask set:
```

```

> SERVO_FTW_MASK = 15

Reviewer 2: param not linked

```

> For other types you will need to calculate it yourself, here is again the example for 4 motors:

Reviewer 2: think this is redundant, I  would remove this...

Reviewer 2: I made example for different setup

```
MOTOR 	    1 | 2 | 3 | 4 | 5 | 6  ...
BIN VALUE   1 | 2 | 4 | 8 | 16| 32 ...
ACTIVE      X | X | X | X | 0 | 0  ...
SUM         1 + 2 + 4 + 8 = 15
```

> Buying

Reviewer 2: Where to Buy

Author: done

- :ref:`SERIAL5_PROTOCOL<SERIAL5_PROTOCOL>` = 38 (FETtec_OneWire)
- :ref:`SERIAL5_OPTIONS<SERIAL5_OPTIONS>` = 0 (default,full duplex), for use with a diode or, :ref:`SERIAL5_OPTIONS<SERIAL5_OPTIONS>` = 4 (half duplex, for use without a diode

This digital protocol can use the full 1000 to 2000 range. Do set these parameters as well:

- :ref:`MOT_PWM_MAX<MOT_PWM_MAX>` = 2000
- :ref:`MOT_PWM_MIN<MOT_PWM_MIN>` = 1000

Change SERVOx_FUNCTION parameters according to the number of motors and their order. e.g. for a quadcopter:

- :ref:`SERVO1_FUNCTION<SERVO1_FUNCTION>` = 33
- :ref:`SERVO2_FUNCTION<SERVO2_FUNCTION>` = 34
- :ref:`SERVO3_FUNCTION<SERVO3_FUNCTION>` = 35
- :ref:`SERVO4_FUNCTION<SERVO4_FUNCTION>` = 36

After that you must set the servo output channels you want to use in the :ref:`SERVO_FTW_MASK<SERVO_FTW_MASK>` parameter, and again restart your autopilot. For example, a quadcopter using the first four motors would need the first four bits of the bitmask set:

:ref:`SERVO_FTW_MASK<SERVO_FTW_MASK>` = 15

For other types you will need to calculate it yourself, here, again,for 4 motors. On servo outputs 1/4/5/6 :

+------------+-+-+-+-+-+-+

> |MOTOR 	     |1|2|3|4|5|6| 

Reviewer 2: somehow this got miss-aligned....I will fix before merge on last test build

===================
FETtec OneWire ESCs
===================

.. note::

   This feature is only available on github master branch and on ArduPilot-4.1 (and higher)

.. image:: ../../../images/FETtec_4in1_ESC.jpg

> OneWire is an `ESC bi-directional communication protocol<https://github.com/ArduPilot/ardupilot/tree/master/libraries/AP_FETtec_OneWire>`_ from FETtec.

Reviewer 2: this is a malformed link..it needs a space in front of the first brace on url..and it should not point to the code, but rather an explanation of the protocol (https://blog.seidel-philipp.de/what-is-fettec-onewire/), or just remove the link entirely

Author: It does not point to code, it points to a README.md file AND the source code! All in one! :)

FETtec OneWire ESC's telemetry information is sent back to the autopilot:

> - Motor rotations per minute (RPM)

Reviewer 3: ```suggestion
- Motor revolutions per minute (RPM)
```

- Input voltage (V)
- Current draw (A)
- Power consumption (W)
- Temperature (°C)
- Summed CRC errors

The autopilot uses it to:

> - log the status of each ESC to the SDCard or internal Flash, for :ref:`post flight analysis<common-logs>`_

Reviewer 2: ```suggestion
- log the status of each ESC to the SDCard or internal Flash, for :ref:`post flight analysis<common-logs>`
```
build error

> - send the status of each ESC to the Ground Station or companion computer for :ref:`real-time monitoring<common-mission-planner-telemetry-logs>`_

Reviewer 2: ```suggestion
- send the status of each ESC to the Ground Station or companion computer for :ref:`real-time monitoring<common-mission-planner-telemetry-logs>`
```
build error

> - optionaly dynamically change the center frequency of the notch filters used to reduce frame vibration noise in the gyros. See optional configuration information bellow.

Reviewer 3: ```suggestion
- optionally dynamically change the center frequency of the notch filters used to reduce frame vibration noise in the gyros. See optional configuration information bellow.
```

> - optionaly measure battery voltage and power consumption. See optional configuration information bellow.

Reviewer 3: ```suggestion
- optionally measure battery voltage and power consumption. See optional configuration information bellow.
```

Where to buy
============

Multiple OneWire supported ESCs are also available at `FETtec <https://fettec.net>`_

Connecting
==========

The protocol supports up-to 25 ESCs, but ArduPilot's device driver implementation limits this to 12.

> Please make sure, that your ESC Firmware Version is 193 (Juli 2021) or later. Older versions only support half-duplex OneWire.

Reviewer 3: ```suggestion
Please make sure, that your ESC Firmware Version is 193 (July 2021) or later. Older versions only support half-duplex OneWire.
```
will be released in July?

This digital protocol can use the full 1000 to 2000 range. Do set these parameters as well:

- :ref:`MOT_PWM_MAX<MOT_PWM_MAX>` = 2000
- :ref:`MOT_PWM_MIN<MOT_PWM_MIN>` = 1000

Change SERVOx_FUNCTION parameters according to the number of motors and their order. e.g. for a quadcopter:

- :ref:`SERVO1_FUNCTION<SERVO1_FUNCTION>` = 33
- :ref:`SERVO2_FUNCTION<SERVO2_FUNCTION>` = 34
- :ref:`SERVO3_FUNCTION<SERVO3_FUNCTION>` = 35
- :ref:`SERVO4_FUNCTION<SERVO4_FUNCTION>` = 36

FTW Mask:
After that you must set the servo output channels you want to use in the :ref:`SERVO_FTW_MASK<SERVO_FTW_MASK>` parameter, and again restart your autopilot. For example, a quadcopter using the first four motors would need the first four bits of the bitmask set:

:ref:`SERVO_FTW_MASK<SERVO_FTW_MASK>` = 15

Each bit of the mask describes whether a motor is attached and should be used with oneWire. The value is configured decimal and is the sum of the decimal equivalent of each set bit. Bit 1 corresponds to 2^1 = 1, Bit 2 corresponds to 2^2 = 2,... . E.g. if you have connected Motor 1,2,3,4 the mask is 1+2+4+8 = 15.  

Simply sum the decimal values for the motors connected or read it from the list below.

> +---+----+

Reviewer 2: malformed table

Reviewer 2: must make all rows the same form and width as the Motor Count | Mask row...all "|" and "+" must be aligned or the table will not display

|Motor Count | Mask|
+---+----+
|1  |1   |
+---+----+
|2  |3   |
+---+----+
|3  |7   |
+---+----+
|4  |15  |
+---+----+
|5  |31  |
+---+----+
|6  |63  |
+---+----+
|7  |127 |
+---+----+
|8  |255 |
+---+----+
|9  |511 |
+---+----+
|10 |1023|
+---+----+
|11 |2047|
+---+----+
|12 |4095|
+---+----+

Optional features can be configured using the instructions at their respective wiki page:

> - dynamically change the center frequency of the :ref:`notch filters used to reduce frame vibration noise in the gyros<common-imu-notch-filtering#dynamic-harmonic-notch>`_

Reviewer 2: ```suggestion
- dynamically change the center frequency of the :ref:`notch filters used to reduce frame vibration noise in the gyros<common-imu-notch-filtering#dynamic-harmonic-notch>`
```
build error

> - measure :ref:`battery voltage, current and power consumption<common-dshot-blheli32-telemetry#use-as-battery-monitor>`_

Reviewer 2: ```suggestion
- measure :ref:`battery voltage, current and power consumption<common-dshot-blheli32-telemetry#use-as-battery-monitor>`
```
build error

> The :ref:`FETtec ESC configurator<https://github.com/FETtec/ESC-Configurator/releases>`_ can change all ESC's settings and do ESC firmware updates.

Reviewer 2: ```suggestion
The :ref:`FETtec ESC configurator<https://github.com/FETtec/ESC-Configurator/releases>` can change all ESC's settings and do ESC firmware updates.
```
build error

These are also made via OneWire so a passthough is possible.
Often used parameters are for example: 

- **Motor direction** - If your motor spins in the wrong direction you can change it easily without rewiring
- **Motor beeps** - Enables or disables motor beeps
- **Soft brake** - If you have props that "unscrew" if they are stopped too fast you can use the softbreak option. 

Troubleshooting
===============

- ESCs keep initializing
   If the flight-control is not receiving telemetry signals from every ESC, it reinitializes the ESCs every five seconds until every ESC is configured correctly.

>    The FTW mask is used to determin the motor count that must be found. Double check if the FTW-Mask is calculated correctly to fit the actual motor count.

Reviewer 3: ```suggestion
   The FTW mask is used to determine the motor count that must be found. Double check if the FTW-Mask is calculated correctly to fit the actual motor count.
```

===================
FETtec OneWire ESCs
===================

.. note::

>    This feature is only available on github master branch and on ArduPilot-4.1 (and higher)

Reviewer 2: ```suggestion
   This feature is available on firmware revisions 4.1 and higher
```

+------------+----+
|Motor Count |Mask|
+------------+----+
|1           |1   |
+------------+----+
|2           |3   |
+------------+----+
|3           |7   |
+------------+----+
|4           |15  |
+------------+----+
|5           |31  |
+------------+----+
|6           |63  |
+------------+----+
|7           |127 |
+------------+----+
|8           |255 |
+------------+----+
|9           |511 |
+------------+----+
|10          |1023|
+------------+----+
|11          |2047|
+------------+----+
|12          |4095|
+------------+----+

FTW RVMask:
After that you must set the servo output channels motor rotation direction in the :ref:`SERVO_FTW_RVMASK<SERVO_FTW_RVMASK>` parameter. To leave to motors with their default rotation direction do:

:ref:`SERVO_FTW_RVMASK<SERVO_FTW_RVMASK>` = 0

Each bit of the mask describes whether a motor is rotation direction should be reversed. The value is configured decimal and is the sum of the decimal equivalent of each set bit. Bit 1 corresponds to 2^1 = 1, Bit 2 corresponds to 2^2 = 2,... .  

Optional features can be configured using the instructions at their respective wiki page:

> - dynamically change the center frequency of the :ref:`notch filters used to reduce frame vibration noise in the gyros<common-imu-notch-filtering#dynamic-harmonic-notch>`

Reviewer 2: incorrect link.... you would need to add a tag to that paragraph and reference it directly..you can reference it indirectly by saying "see Dynamic Harmonics section of" and then put the link to the common-imu-notch-filtering  page without the #subdirect

Author: I used https://sublime-and-sphinx-guide.readthedocs.io/en/latest/references.html#use-custom-link-text to create the links. I hope that is OK now

> - measure :ref:`battery voltage, current and power consumption<common-dshot-blheli32-telemetry#use-as-battery-monitor>`

Reviewer 2: same thing here

> The :ref:`FETtec ESC configurator <https://github.com/FETtec/ESC-Configurator/releases>` can change all ESC's settings and do ESC firmware updates.

Reviewer 2: this link is malformed....use a double underscore immediately following the backtick

> FETtec OneWire

Reviewer 2: need equal signs above this header also

==============

This is a half duplex serial protocol with a 2Mbit/s Baudrate created by Felix Niessen (former Flyduino KISS developer) from `FETtec <https://fettec.net/>`.
For example, you can connect a 4in1 ESC with only 1 signal wire and 1 GND wire to your flight controller.
The OneWire protocol is limited to 24 ESCs per OneWire bus.

> You can change way more settings of the ESC over OneWire than over DShot.

Reviewer 2: ```suggestion
You can change many more settings of the ESC over OneWire than over DShot.
```

It also has a better cyclic redundancy check (CRC) than Dshot offers.
CRC checks if there are any signal errors that occur in the communication between autopilot and ESC; and between ESC and autopilot.
It supports ESC telemetry, so information from the ESC status can be sent back to the autopilot:

- Motor rotations per minute (RPM)
- Input voltage (V)
- Current draw (A)
- Power consumption (W)
- Temperature (°C)

This information allows the autopilot to:

- dynamicaly change the center frequency of the notch filters used to reduce frame vibration noise in the gyros
- log the status of each ESC to the SDCard or internal Flash, for post flight analysis
- send the status of each ESC to the Ground Station or companion computer for real-time monitoring

Connecting
==========

The protocol supports up-to 25 ESCs, but the ArduPilot implementation of it limits this to 12.

> You can connect all ESCs to a single serial port using only a single signal wire (TX) and one GND wire.

Reviewer 2: ```suggestion
You can connect all ESCs to a single serial port using only a single signal wire from the autopilot's UART TX pin to the ESC's TLM pin and one GND wire.
```

If you are using 4in1 ESCs you can connect up-to 3 of them in this fashion.

Configuring
===========

To configure it you need to set the SERIALx_PROTOCOL parameter to FETtec_OneWire (37) and SERIALx_OPTIONS parameter to HalfDuplex (4), where x is the number of the serial port you are using.

For example:

> SERIAL2_PROTOCOL = 37

Reviewer 2: link params using :ref:backtick....formats...cant post here because MD mis-interprets the backtick

Author: Please do try to post because I can not find any working link to copy from. The ones I find have nice formatting, but do not link to anything :(

SERIAL2_OPTIONS = 4

> Then you need to reboot the autopilot. After that you must set the channels you want to use in the SERVO_FTW_MASK parameter

Reviewer 2: this param also

Author: SERVO_FTW_MASK Is not in any generated page yet, so I can not link to it`s documentation until the ardupilot code PR gets merged

Reviewer 2: you can create the link to what it WILL be though....

, and again restart your autopilot.

For example, a quadcopter using the first four motors would need
SERVO_FTW_MASK = 15

Buying
======

> Multiple ESCs are available at <https://fettec.net>

Reviewer 2: unfortunately cant show proper format
```suggestion
Multiple ESCs are available at (backtick) FETtec <https://fettec.net>(backtick)( underscore)
```

==============
FETtec OneWire
==============

This is a half duplex serial protocol with a 2Mbit/s Baudrate created by Felix Niessen (former Flyduino KISS developer) from `FETtec <https://fettec.net/>`.

> You can change many more settings of the ESC over OneWire than over DShot.

Reviewer: This is advertisement statement that should be removed

Author: @teezeefpv can you provide proff of this? Or is it just mambo-jumbo?
Originaly I got this from https://blog.seidel-philipp.de/what-is-fettec-onewire/ because that page actualy has more information than your website. :(

Reviewer: This bring nothing on ArduPilot, we should just remove it. And let people do their search on what protocol they think are the better.

Author: Done

> It also has a better cyclic redundancy check (CRC) than Dshot offers.

Reviewer: Same

Author: @teezeefpv again, the same as above.

===================
FETtec OneWire ESCs
===================

> .. note::

Reviewer 2: we dont do this...if it will be in master only, we will just add this page to Upcoming features....

Author: I hope this will be released at some point, and users of 4.0 should not be surprised if this does not work for them

   This feature is available in Copter-4.1 (and higher)

> OneWire is an `ESC bi-directional communication protocol<https://github.com/ArduPilot/ardupilot/tree/master/libraries/AP_FETtec_OneWire>`_ from `FETtec <https://fettec.net>`_.

Reviewer 2: having the url twice on a page causes a build error....that is why I removed it from this sentence

Author: Why is that a build error ??

Reviewer 2: IDK, but it is...Sphinx is funny about some things

This information allows the autopilot to:

- log the status of each ESC to the SDCard or internal Flash, for `post flight analysis<https://ardupilot.org/copter/docs/common-logs.html>`_
- send the status of each ESC to the Ground Station or companion computer for `real-time monitoring<https://ardupilot.org/copter/docs/common-logs.html>`_

> - optionaly dynamically change the center frequency of the `notch filters used to reduce frame vibration noise in the gyros<https://ardupilot.org/copter/docs/common-imu-notch-filtering.html#dynamic-harmonic-notch>`_

Reviewer 2: avoid https links to wiki pages...use the page tag

Author: I think I now have this, but how do I link to an html anchor?

==========

The protocol supports up-to 25 ESCs, but the ArduPilot implementation limits this to 12.
You must connect all ESCs to a single serial port using a single signal wire from the autopilot's UART TX pin to the ESC's TLM pin and one GND wire.
If you are using 4in1 ESCs you can connect up-to 3 of them in this fashion.

TODO: @teezeefpv add a figure here.

FullDuplex with Diode
=====================

> If halfduplex is not available or working, there is a workaround with a diode using the normal full duplex.

Reviewer 2: i fyou are trying for a line break, this does not work..must have a blank line,else it will be concatenated onto the previous line....you can force single space by using sphinx block markers though

Author: No, I was just making it neater

:ref:`SERVO_FTW_MASK<SERVO_FTW_MASK>` = bit0/2/3/5 set = 1+4+8+32= 45

> Optional features can be configured using the instructions at their respective wiki page:

Reviewer 2: same comment as above...use page tags not https
